src/components/Library.vue

```diff
@@ -175,52 +180,85 @@ export default {
+            break;
+        }
       }
       this.selectLatest();
     },
     async duplicateRow(row) {
-      this.createObject();
-      await this.$nextTick();
-      let newName = row.name;
-      // increment Copy # until no name conflict.
-      while (_.find(this.rows, { name: newName })) {
-        let copyNum = newName.match(/\((\d+)\)/);
-        newName = copyNum ? `${newName.slice(0, copyNum.index)}(${+copyNum[1] + 1})` : `${newName} (2)`;
+      if(this.selectedObject.name.includes('Story')) {
+        this.cloneStory(this.selectedObject);
+      } else {
+        this.createObject();
+        await this.$nextTick();
+        let newName = row.name;
+        // increment Copy # until no name conflict.
+        while (_.find(this.rows, { name: newName })) {
+          let copyNum = newName.match(/\((\d+)\)/);
+          newName = copyNum ? `${newName.slice(0, copyNum.index)}(${+copyNum[1] + 1})` : `${newName} (2)`;
+        }
+        this.modifyObject(this.selectedObject.id, 'name', newName);
+        this.columns.forEach(({ name: key, readonly, private: privateKey }) => {
+          if (key === 'id' || key === 'name' || key === 'color') return;
+          if (readonly || privateKey) return;
+          if (this.selectedObject[key] === row[key]) return;
+          this.modifyObject(this.selectedObject.id, key, row[key]);
+        });
       }
-      this.modifyObject(this.selectedObject.id, 'name', newName);
-      this.columns.forEach(({ name: key, readonly, private: privateKey }) => {
-        if (key === 'id' || key === 'name' || key === 'color') return;
-        if (readonly || privateKey) return;
-        if (this.selectedObject[key] === row[key]) return;
-        this.modifyObject(this.selectedObject.id, key, row[key]);
-      });
     },
     selectLatest() {
       const newestRow = _.maxBy(this.rows, r => +r.id);
       if (!newestRow) { return; }
       this.selectedObject = newestRow;
     },
+    cloneStory(story) {
+      this.$store.dispatch('application/setCurrentStoryId', { id: story.id });
+      const { clonedGeometry, idMap } = replaceIdsForCloning(this.currentStoryGeom);
+      this.createObject({ duplicate: true });
+      const { clonedStory } = modelHelpers.replaceIdsUpdateInfoForCloning(story, idMap, this.state, this.currentStory);
+      this.destroyDuplicateSpaces('spaces', this.currentStory.spaces[0]);
+      this.$store.dispatch('models/cloneStory', clonedStory);
+      this.$store.dispatch('geometry/cloneStoryGeometry', clonedGeometry);
+      for (let i = this.currentStory.shading.length - 1; i >= 0; i--) {
+        this.$store.dispatch('models/destroyShading', {
+          shading: this.currentStory.shading[i],
+          story: this.currentStory,
+        });
+      }
+    },
+    destroyDuplicateSpaces(type, object) {
+      this.$store.dispatch('models/destroySpace', {
+        space: object,
+        story: this.$store.state.models.stories.find(story => story[type].find(o => o.id === object.id)),
```

Comment:
Could pass in `this.currentStory` instead

Reply:
Implemented.